### ISSUE TITLE
Bug: [consistent-indexed-object-style] Fails on generic circular reference

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#5472]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=KYDwDg9gTgLgBASwHY2FAZgQwMbDgGQQCMpMoBPOAbwCg44BtBALjgGcYpkBzAXVcIkylAD7tOPBrwDcNAL40aoSLEQo0WXAWKkKAcWBI0CbAB4AKgD5qdRi3FckfVubhjBu8gaNczV2Qo0QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6MgeyeUuX0Ra1mAE0QAPRMNocARgCtEZOn0JJ0URNGgdokcGAC+IA0A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
export interface Library {
  [i: string]: Library | string[];
}

export interface LibraryGeneric<T> {
  [i: string]: T | LibraryGeneric<T>;
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/consistent-indexed-object-style": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

`interface LibraryGeneric` can't be represented by a `type` for the same reason `Library` can't, per #4347. The rule shouldn't report on it.

### Actual Result

The rule reports on `LibraryGeneric`.

### Additional Info

Like #2687, but with generics.

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.33.0` |
| `@typescript-eslint/parser` | `5.33.0` |
| `TypeScript` | `4.7.2` |
| `ESLint` | `8.15.0` |
| `node` | `web` |
